File: src/core/SaveData.js
import { Emitter } from './Util.js';

const KEY = 'rotgrid.save.v1';

const EMPTY = {
  lifetimeKills: 0,
  lifetimeHeadshots: 0,
  highestRound: 0,
  gamesPlayed: 0,
  totalTimeSurvived: 0,
  totalPointsEarned: 0,
  totalDowns: 0,
  totalRevives: 0,
  barriersRepaired: 0,
  boxSpins: 0,
  weaponStats: {},   // id -> { kills, headshots, shotsFired, shotsHit, damage }
  perkPurchases: {}, // id -> count
  history: [],       // last 10 runs
};

class SaveStore extends Emitter {
  constructor() {
    super();
    this.data = structuredClone(EMPTY);
    this.load();
  }

  load() {
    try {
      const raw = localStorage.getItem(KEY);
      if (raw) this.data = { ...structuredClone(EMPTY), ...JSON.parse(raw) };
    } catch { /* ignore */ }
  }

  save() {
    try { localStorage.setItem(KEY, JSON.stringify(this.data)); } catch { /* ignore */ }
  }

  weapon(id) {
    if (!this.data.weaponStats[id]) {
      this.data.weaponStats[id] = { kills: 0, headshots: 0, shotsFired: 0, shotsHit: 0, damage: 0 };
    }
    return this.data.weaponStats[id];
  }

  /** Fold a completed run into lifetime totals. */
  recordRun(run) {
    const d = this.data;
    d.gamesPlayed++;
    d.lifetimeKills += run.kills;
    d.lifetimeHeadshots += run.headshots;
    d.totalTimeSurvived += run.time;
    d.totalPointsEarned += run.pointsEarned;
    d.totalDowns += run.downs;
    d.totalRevives += run.revives;
    d.barriersRepaired += run.repairs;
    d.boxSpins += run.boxSpins;
    const newBest = run.round > d.highestRound;
    if (newBest) d.highestRound = run.round;

    for (const [id, s] of Object.entries(run.weaponStats || {})) {
      const w = this.weapon(id);
      w.kills += s.kills || 0;
      w.headshots += s.headshots || 0;
      w.shotsFired += s.shotsFired || 0;
      w.shotsHit += s.shotsHit || 0;
      w.damage += s.damage || 0;
    }
    for (const id of run.perksBought || []) {
      d.perkPurchases[id] = (d.perkPurchases[id] || 0) + 1;
    }

    d.history.unshift({
      date: Date.now(), round: run.round, kills: run.kills,
      headshots: run.headshots, time: run.time, points: run.pointsEarned,
    });
    d.history = d.history.slice(0, 10);

    this.save();
    this.emit('update', d);
    return newBest;
  }

  reset() {
    this.data = structuredClone(EMPTY);
    this.save();
    this.emit('update', this.data);
  }
}

export const SaveData = new SaveStore();

File: unity/Rotgrid/Assets/Scripts/Core/SaveData.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace Rotgrid
{
    [Serializable]
    public class WeaponRecord
    {
        public string id;
        public int kills;
        public int headshots;
        public int shotsFired;
        public int shotsHit;
        public int damage;
    }

    [Serializable]
    public class RunRecord
    {
        public string date;
        public int round;
        public int kills;
        public int headshots;
        public float time;
        public int points;
    }

    [Serializable]
    public class CareerData
    {
        public int lifetimeKills;
        public int lifetimeHeadshots;
        public int highestRound;
        public int gamesPlayed;
        public float totalTimeSurvived;
        public int totalPointsEarned;
        public int totalDowns;
        public int totalRevives;
        public int barriersRepaired;
        public int boxSpins;
        public List<WeaponRecord> weapons = new List<WeaponRecord>();
        public List<RunRecord> history = new List<RunRecord>();
    }

    /// <summary>Lifetime statistics, persisted as JSON in PlayerPrefs.</summary>
    public static class SaveData
    {
        const string Key = GameSettings.PrefPrefix + "career";
        public static CareerData Data = new CareerData();

        public static void Load()
        {
            string json = PlayerPrefs.GetString(Key, "");
            if (string.IsNullOrEmpty(json)) { Data = new CareerData(); return; }
            try
            {
                var parsed = JsonUtility.FromJson<CareerData>(json);
                Data = parsed ?? new CareerData();
            }
            catch (Exception) { Data = new CareerData(); }
            if (Data.weapons == null) Data.weapons = new List<WeaponRecord>();
            if (Data.history == null) Data.history = new List<RunRecord>();
        }

        public static void Save()
        {
            PlayerPrefs.SetString(Key, JsonUtility.ToJson(Data));
            PlayerPrefs.Save();
        }

        public static WeaponRecord Weapon(string id)
        {
            for (int i = 0; i < Data.weapons.Count; i++) if (Data.weapons[i].id == id) return Data.weapons[i];
            var rec = new WeaponRecord { id = id };
            Data.weapons.Add(rec);
            return rec;
        }

        /// <summary>Folds a finished run into the career totals. Returns true on a new best.</summary>
        public static bool RecordRun(RunSummary s)
        {
            Data.gamesPlayed++;
            Data.lifetimeKills += s.kills;
            Data.lifetimeHeadshots += s.headshots;
            Data.totalTimeSurvived += s.time;
            Data.totalPointsEarned += s.pointsEarned;
            Data.totalDowns += s.downs;
            Data.totalRevives += s.revives;
            Data.barriersRepaired += s.repairs;
            Data.boxSpins += s.boxSpins;

            bool newBest = s.round > Data.highestRound;
            if (newBest) Data.highestRound = s.round;

            foreach (var kv in s.weaponStats)
            {
                var rec = Weapon(kv.Key);
                rec.kills += kv.Value.kills;
                rec.headshots += kv.Value.headshots;
                rec.shotsFired += kv.Value.shotsFired;
                rec.shotsHit += kv.Value.shotsHit;
                rec.damage += kv.Value.damage;
            }

            Data.history.Insert(0, new RunRecord
            {
                date = DateTime.Now.ToString("yyyy-MM-dd"),
                round = s.round,
                kills = s.kills,
                headshots = s.headshots,
                time = s.time,
                points = s.pointsEarned,
            });
            while (Data.history.Count > 10) Data.history.RemoveAt(Data.history.Count - 1);

            Save();
            return newBest;
        }

        public static void Wipe()
        {
            Data = new CareerData();
            Save();
        }
    }

    /// <summary>Per-run tally handed to the death screen and the career store.</summary>
    public class RunSummary
    {
        public int round;
        public int roundsSurvived;
        public int kills;
        public int headshots;
        public int meleeKills;
        public int repairs;
        public float time;
        public int pointsEarned;
        public int shotsFired;
        public int shotsHit;
        public int downs;
        public int revives;
        public int boxSpins;
        public bool newBest;
        public List<string> perksBought = new List<string>();
        public Dictionary<string, WeaponRecord> weaponStats = new Dictionary<string, WeaponRecord>();

        public WeaponRecord Weapon(string id)
        {
            WeaponRecord r;
            if (!weaponStats.TryGetValue(id, out r))
            {
                r = new WeaponRecord { id = id };
                weaponStats[id] = r;
            }
            return r;
        }
    }
}
